GiniVision/Classes/Core/Screens/Multipage Review/MultipageReviewCollectionsCellPresenter.swift

```diff
@@ -0,0 +1,209 @@
+final class MultipageReviewCollectionCellPresenter {
+    
+    weak var delegate: MultipageReviewCollectionCellPresenterDelegate?
+    var thumbnails: [String: [ThumbnailType: UIImage]] = [:]
+    fileprivate let giniConfiguration: GiniConfiguration
+    fileprivate let thumbnailsQueue = DispatchQueue(label: "Thumbnails queue")
+    
+    enum MultipageCollectionCellType {
+        case main(MultipageReviewMainCollectionCell, (NoticeActionType) -> Void)
+        case pages(MultipageReviewPagesCollectionCell)
+    }
+    
+    enum ThumbnailType {
+        case big, small
+        
+        var scale: CGFloat {
+            switch self {
+            case .big:
+                return 1.0
+            case .small:
+                return 1/4
+            }
+        }
+    }
+    
+    init(giniConfiguration: GiniConfiguration = .shared) {
+        self.giniConfiguration = giniConfiguration
+    }
+    
+    func setUp(_ cell: MultipageCollectionCellType,
+               with page: GiniVisionPage,
+               isSelected: Bool,
+               at indexPath: IndexPath) -> UICollectionViewCell {
+        
+        let collectionCell: UICollectionViewCell
+        switch cell {
+        case .main(let mainCell, let errorAction):
+            setUp(cell: mainCell,
+                  with: page,
+                  fetchThumbnailTrigger: fetchThumbnailImage(for: page, of: .big, in: cell, at: indexPath),
+                  didTapErrorNotice: errorAction)
+            collectionCell = mainCell
+        case .pages(let pageCell):
+            setUp(cell: pageCell,
+                  with: page,
+                  fetchThumbnailTrigger: fetchThumbnailImage(for: page, of: .small, in: cell, at: indexPath),
+                  pageIndex: indexPath.row,
+                  selected: isSelected)
+            collectionCell = pageCell
+        }
+        
+        return collectionCell
+    }
+    
+    func rotateThumbnails(for page: GiniVisionPage) {
+        if let smallRotatedImage = thumbnails[page.document.id]?[.small]?.rotated90Degrees() {
+            thumbnails[page.document.id]![.small] = smallRotatedImage
+        }
+        
+        if let bigRotatedImage = thumbnails[page.document.id]?[.big]?.rotated90Degrees() {
+            thumbnails[page.document.id]![.big] = bigRotatedImage
+        }
+    }
+}
+
+// MARK: - Cells setup
+
+fileprivate extension MultipageReviewCollectionCellPresenter {
+    
+    // MARK: - MultipageReviewPagesCollectionCell
+    
+    func setUp(cell: MultipageReviewPagesCollectionCell,
+               with page: GiniVisionPage,
+               fetchThumbnailTrigger: @autoclosure () -> Void,
+               pageIndex: Int,
+               selected: Bool) {
+        // Thumbnail set
+        if let thumbnail = self.thumbnails[page.document.id, default: [:]][.small] {
+            cell.documentImage.contentMode = thumbnail.size.width > thumbnail.size.height ?
+                .scaleAspectFit :
+                .scaleAspectFill
```

Comment:
Doesn't `scaleAspectFit` work well for both kinds of ratio?

Reply:
It doesn't look very well, mostly if the ratio width/height is too low. That's why is preferred to use AspectFill for portrait images.